File: LinkedLists/C++/reverse.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

Node* head;

void reverseList() {
    Node* current = head;
    Node* prev = NULL;
    Node* next = NULL;
    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}

void testReverseList() {
    // Create a linked list with 3 nodes
    Node* node1 = new Node();
    node1->data = 1;
    head = node1;
    Node* node2 = new Node();
    node2->data = 2;
    node1->next = node2;
    Node* node3 = new Node();
    node3->data = 3;
    node2->next = node3;
    node3->next = NULL;

    // Print the original list
    cout << "Original List: ";
    Node* current = head;
    while (current != NULL) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;

    // Reverse the list
    reverseList();

    // Print the reversed list
    cout << "Reversed List: ";
    current = head;
    while (current != NULL) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;
}

int main() {
    testReverseList();
    return 0;
}

File: LinkedLists/Python/reverse.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def reverse(self):
        current = self.head
        prev = None
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        self.head = prev

def test_reverse():
    # Create a linked list with 3 nodes
    ll = LinkedList()
    ll.head = Node(1)
    second = Node(2)
    third = Node(3)
    ll.head.next = second
    second.next = third

    # Print the original list
    current = ll.head
    print("Original List: ", end='')
    while current:
        print(current.data, end=' ')
        current = current.next
    print()

    # Reverse the list
    ll.reverse()

    # Print the reversed list
    current = ll.head
    print("Reversed List: ", end='')
    while current:
        print(current.data, end=' ')
        current = current.next
    print()

test_reverse()
